dll/win32/kernel32/kernel32_vista/GetFinalPathNameByHandle.c

```diff
@@ -0,0 +1,298 @@
+                          FILE_OPEN_IF,
+                          0,
+                          NULL,
+                          0);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("NtCreateFile failed: %lx\n", Status);
+        return Status;
+    }
+
+    /* Use the MountMgr to query the volume name */
+    TargetName->DeviceNameLength = DeviceName->Length;
+    RtlCopyMemory(TargetName->DeviceName, DeviceName->Buffer, DeviceName->Length);
+    Status = NtDeviceIoControlFile(MountMgrHandle,
+                                   NULL,
+                                   NULL,
+                                   NULL,
+                                   &IoStatusBlock,
+                                   IOCTL_MOUNTMGR_QUERY_DOS_VOLUME_PATH,
+                                   Buffer,
+                                   sizeof(Buffer),
+                                   Buffer,
+                                   sizeof(Buffer));
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("NtDeviceIoControlFile failed: %lx\n", Status);
+        goto Exit;
+    }
+
+    if ((IoStatusBlock.Information < sizeof(*VolumePaths)) ||
+        (IoStatusBlock.Information < (VolumePaths->MultiSzLength + sizeof(ULONG))))
+    {
+        DPRINT1("Invalid information returned: %lu\n", IoStatusBlock.Information);
+        ASSERT(FALSE);
+        Status = STATUS_UNSUCCESSFUL;
+        goto Exit;
+    }
+
+    /* Construct the full volume path name from the first returned DOS volume */
+    if (!NT_SUCCESS(RtlAppendUnicodeToString(DosVolumeName, L"\\\\?\\")) ||
+        !NT_SUCCESS(RtlAppendUnicodeToString(DosVolumeName, VolumePaths->MultiSz) ||
+        !NT_SUCCESS(RtlAppendUnicodeToString(DosVolumeName, L"\\"))))
+    {
+        DPRINT1("RtlAppendUnicodeToString failed: %lx\n", Status);
+        goto Exit;
+    }
+
+    /* This is based on Windows behavior */
+    SetLastError(ERROR_SUCCESS);
+
+Exit:
+    NtClose(MountMgrHandle);
+
+    return Status;
+}
+
+DWORD
+WINAPI
+GetFinalPathNameByHandleW(
+    _In_ HANDLE hFile,
+    _Out_writes_(cchFilePath) LPWSTR lpszFilePath,
+    _In_ DWORD cchFilePath,
+    _In_ DWORD dwFlags)
+{
+    WCHAR NameInfoBuffer[sizeof(OBJECT_NAME_INFORMATION) + MAX_PATH + 1];
+    POBJECT_NAME_INFORMATION NameInfo = (POBJECT_NAME_INFORMATION)NameInfoBuffer;
+    PFILE_NAME_INFORMATION FileNameInfo = (PFILE_NAME_INFORMATION)NameInfoBuffer;
+    WCHAR VolumeBuffer[sizeof("\\Device\\Volume{01234567-89ab-cdef-0123-456789abcdef}")];
+    IO_STATUS_BLOCK IoStatusBlock;
+    PWSTR PathSplit;
+    UNICODE_STRING DeviceName, RelativePath, VolumeName;
+    SIZE_T VolumeLength;
+    ULONG VolumeType;
+    NTSTATUS Status;
+    ULONG FinalLength;
+    BOOL Success;
+
+    /* Extract the requested volume type */
+    VolumeType = dwFlags & (VOLUME_NAME_GUID | VOLUME_NAME_NONE | VOLUME_NAME_NT);
+
+    /* Check if more than one volume type flag is set */
+    if (VolumeType & (VolumeType- 1))
```

Comment:
```suggestion
    if (VolumeType & (VolumeType - 1))
```

Reply:
The function doesn't verify if other unknown flags are passed?
For example in the apitest, this line:
```c
pGetFinalPathNameByHandleW(hFile, Buffer, ARRAYSIZE(Buffer), VOLUME_NAME_GUID | FILE_NAME_OPENED);
```
which appears to succeed: what happens if you add `0xFFFFFFF0` to the flags?